.gitlab-ci.yml

```diff
@@ -1,5 +1,47 @@
+.tmpl:master_rules: &tmpl_master_rules
+  rules:
+    - if: $CI_COMMIT_BRANCH == 'master'
+      when: always
+  stage: deploy
+  variables:
+    SLS_STAGE: production
+
+.tmpl:branch_rules: &tmpl_branch_rules
+  rules:
+    - if: $CI_EXTERNAL_PULL_REQUEST_IID
+      when: always
+  stage: deploy
+  variables:
+    SLS_STAGE: $CI_COMMIT_REF_SLUG
+
+.tmpl:branch_remove_rules: &tmpl_branch_remove_rules
+  rules:
+    - if: $CI_EXTERNAL_PULL_REQUEST_IID
+      when: manual
+      allow_failure: true
+  stage: deploy
+  variables:
+    GIT_CHECKOUT: 0
+    SLS_STAGE: $CI_COMMIT_REF_SLUG
+
+.tmpl:deploy: &tmpl_deploy
+  artifacts:
+    reports:
+      dotenv: deploy.env
+  script:
+    - yarn pnpify serverless deploy --verbose
+    - echo "APP_URL=$(yarn pnpify serverless info --verbose | grep -w "CloudFrontDistributionDomain" | cut -d ' ' -f2)" >> deploy.env
+  dependencies:
```

Comment:
We need to distinguish between master and branch here. master also needs to depend on all checks.

Reply:
"Dependencies" refer to the artifacts copied to the working directory, "needs" allow for "parallelization". For deploy, we "need" every previous stage to finish.

Reply:
IMO it should be `needs` for branches and `dependencies` for master?
Why should we prevent failing lints, tests etc. from deploying a feature env - e.g. if I want to show someone my changes despite not having fixed all tests yet.

Reply:
We can add needs to branches to avoid waiting for checks to speed up the process but the pipeline will still fail. Is this what you have in mind?

Reply:
Yes, the pipeline should def fail if checks or tests or anything are red - but the deploy can go ahead regardless (on branches only, not master obviously)

Reply:
https://github.com/yldio/asap-hub/pull/4/files#diff-96edf7a6f008de9e928d04e1ae5e12a5R117